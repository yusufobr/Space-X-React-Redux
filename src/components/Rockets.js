import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { addReserve } from '../redux/rockets/rocketsSlice';

function Rockets() {
  const { rockets } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  return (
    <div className="container px-4 mx-auto flex flex-col gap-6 mt-8">
      {rockets.map((rocket) => (
        <div key={rocket.rocket_id} className="flex gap-6">
          <img width={250} src={rocket.flickr_images} alt="flickr" />
          <div className="text-left">
            <h3 className="text-lg font-semibold mb-2">{rocket.rocket_name}</h3>
            <p className="mb-4">
              {
                rocket.reserved ? (
                  <span className="bg-[#18A2B8] mr-2 text-white text-xs font-semibold capitalize px-2 rounded-md">
                    Reserved
                  </span>
                ) : ''
              }
              {rocket.description}
            </p>
            <button
              type="button"
              className={
                    rocket.reserved
                      ? 'py-1 px-3 rounded border-2 text-gray-500 border-gray-500'
                      : 'py-2 px-3 rounded text-white bg-[#007bff]'
                  }
              id={rocket.rocket_id}
              onClick={() => {
                dispatch(addReserve(rocket.rocket_id));
              }}
            >
              {rocket.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
            </button>
          </div>
        </div>
      ))}
      ;
    </div>
  );
}

export default Rockets;
