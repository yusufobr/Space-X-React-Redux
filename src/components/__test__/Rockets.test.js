import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Rockets from '../Rockets';

const mockStore = configureStore([]);
const initialState = {
  rockets: {
    rockets: [
      {
        rocket_id: 'rocket1',
        rocket_name: 'Falcon 1',
        description: 'Description 1',
        flickr_images: 'https://example.com/rocket1.jpg',
        reserved: false,
      },
      {
        rocket_id: 'rocket2',
        rocket_name: 'Falcon 9',
        description: 'Description 2',
        flickr_images: 'https://example.com/rocket2.jpg',
        reserved: true,
      },
    ],
  },
};
const store = mockStore(initialState);

describe('Rockets Component', () => {
  test('renders a list of rockets with reservation functionality', () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    const rocketNames = screen.getAllByText(/Falcon/i);
    expect(rocketNames).toHaveLength(2);

    const reserveButton = screen.getByText(/Reserve Rocket/i);
    fireEvent.click(reserveButton);
    expect(reserveButton.textContent).toBe('Reserve Rocket');

    const reservationStatus = screen.getByText(/Reserved/i);
    expect(reservationStatus).toBeInTheDocument();

    // snapshot testing
    expect('Rockets component').toMatchSnapshot();
  });
});
