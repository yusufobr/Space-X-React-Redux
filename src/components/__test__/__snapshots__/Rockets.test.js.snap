// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Component renders a list of rockets with reservation functionality 1`] = `"Rockets component"`;
